workflow/controller/controller.go

```diff
@@ -542,6 +541,52 @@ func (wfc *WorkflowController) runPodCleanup(ctx context.Context) {
 	}
 }
 
+func (wfc *WorkflowController) getPodFinalizerRemovalPatch(namespace, podName string) (string, bool) {
+	if os.Getenv("ARGO_POD_STATUS_CAPTURE_FINALIZER") != "true" {
+		return "", false
+	}
+	found := false
+	keeps := make([]string, 0)
+	pod, err := wfc.getPod(namespace, podName)
```

Comment:
Is it safe to assume that the Informer cache here is up to date? If other controllers are adding or removing Pod finalizers at the same time, is there a chance that finalizers could be added or removed unintentionally?

Reply:
This is argo workflow-controller managed pod, it seems that no other controller will add finalizers to this pod. If there is such a controller, it should have already added at pod creation time, if other controller is removing the controller, it can remove it one more time when recieving a Update event. But, what else controller wiil add a finalizer to a workflow pod...

Reply:
Ah, so generating an idempotent merge patch can avoid these cases.

Reply:
I followed up on this [below](https://github.com/argoproj/argo-workflows/pull/12862#discussion_r1555109592) (as did Julie in [a comment](https://github.com/argoproj/argo-workflows/pull/12862#issuecomment-2040762523) before that) -- let's continue there.

> But, what else controller wiil add a finalizer to a workflow pod...

There can be many, which is why I brought this part up [below](https://github.com/argoproj/argo-workflows/pull/12862#issuecomment-2028878367) as well. Pods are a generic building block in k8s and are not Argo-specific.

A really simple counter-example is a policy controller that does some clean up. It would run on _all_ Pods, regardless of what controller started them.